compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -340,21 +325,37 @@ public TypeChecker(CompilerContext context) {
         this.anonymousModelHelper = BLangAnonymousModelHelper.getInstance(context);
         this.semanticAnalyzer = SemanticAnalyzer.getInstance(context);
         this.missingNodesHelper = BLangMissingNodesHelper.getInstance(context);
-        this.queryFinalClauses = new Stack<>();
-        this.queryEnvs = new Stack<>();
-        this.prevEnvs = new Stack<>();
         this.unifier = new Unifier();
     }
 
-    public BType checkExpr(BLangExpression expr, SymbolEnv env) {
-        return checkExpr(expr, env, symTable.noType);
+    private BType checkExpr(BLangExpression expr, SymbolEnv env, AnalyzerData data) {
+        return checkExpr(expr, env, symTable.noType, data);
     }
 
-    public BType checkExpr(BLangExpression expr, SymbolEnv env, BType expType) {
-        return checkExpr(expr, env, expType, DiagnosticErrorCode.INCOMPATIBLE_TYPES);
+    private BType checkExpr(BLangExpression expr, SymbolEnv env, BType expType, AnalyzerData data) {
+        return checkExpr(expr, env, expType, DiagnosticErrorCode.INCOMPATIBLE_TYPES, data);
     }
 
-    public BType checkExpr(BLangExpression expr, SymbolEnv env, BType expType, DiagnosticCode diagCode) {
+    public BType checkExpr(BLangExpression expr, SymbolEnv env, BType expType, Stack<SymbolEnv> prevEnvs) {
+        final AnalyzerData data = new AnalyzerData();
+        data.env = env;
+        data.prevEnvs = prevEnvs;
+        data.queryFinalClauses = new Stack<>();
+        data.queryEnvs = new Stack<>();
+        return checkExpr(expr, env, expType, DiagnosticErrorCode.INCOMPATIBLE_TYPES, data);
+    }
```

Comment:
The local variable `env` has been already set to `data`, no need to pass the `env` again

Reply:
Yes. But in other usages of `checkExpr` we need to pass `env` also.